src/coreclr/jit/morph.cpp

```diff
@@ -9869,20 +9869,13 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
 
         ASSIGN_HELPER_FOR_MOD:
 
-            // For "val % 1", return 0 if op1 doesn't have any side effects
-            // and we are not in the CSE phase, we cannot discard 'tree'
-            // because it may contain CSE expressions that we haven't yet examined.
-            //
-            if (((op1->gtFlags & GTF_SIDE_EFFECT) == 0) && !optValnumCSE_phase)
+            if (!optValnumCSE_phase)
             {
-                if (op2->IsIntegralConst(1))
+                if (tree->OperIs(GT_MOD, GT_UMOD) && (op2->IsIntegralConst(1) || op2->IsIntegralConst(-1)))
```

Comment:
Note that `a % -1` throws an exception for `a == int.MinValue`, so I'm not sure the transformation is valid for that case.

Reply:
> Note that `a % -1` throws an exception for `a == int.MinValue`, so I'm not sure the transformation is valid for that case.

Hopefully we have a test for that case somewhere?

Reply:
Ah, here it is:
https://github.com/dotnet/runtime/blob/c6549d9ef9f37539f8ea4c14073e75c7b69fbd23/src/tests/JIT/CodeGenBringUpTests/ModConst.cs#L216

Reply:
Yea I removed the -1 optimization. We probably could optimize it for small types since they don't throw, let's do that in a separate PR.